### DISCUSSION TITLE
Using JSON vs Pickle

### DISCUSSION THREAD

**Discussion by wireboy5** [Slow-Scallion4183/Adventure-Colussus#4] [Show and tell]
**I Put this in show and tell because I do not know which other category it would fit in**
This is just a real quick rundown of JSON and pydantic.

## JSON
JSON (JavaScript-Object-Notation) is a commonly used format for serializing data for storage on a disk, transmission over the network or for storage in a database. There are a few key differences between JSON and pickle.

### JSON is human readable
JSON is designed to be read and written by humans. If you have every used python's dictionary, this should look familiar:
```JSON
{
    "player_name": "Some Name",
    "player_age": 20,
    "health": 100,
    "inventory": [
        {
            "item_name": "sword",
            "item_count": 1
        }
    ]
}
```

Now, pickle can still be decoded by any application, so JSON being human readable makes no difference for security, but can help greatly by speeding up the development process.

### JSON is supported EVERYWHERE
If you wanted to rewrite your game in C++, your players could still use the same characters from the python version.
JSON has a version for almost every programming language in existence.

### JSON is compact
JSON file sizes tend to be smaller than pickle file sizes.

### JSON cannot encode python objects directly
Don't panic. JSON can encode any dictionary or list that only contains dictionaries, lists, strings, and numbers.

In python, everything boils down to these four types. Almost any class can still be encoded, but it requires a method to turn it into a dictionary. This is where pydantic comes in.

## Pydantic
Pydantic can be installed via pip using this command:
```bash
pip install pydantic
```
Pydantic is a python library for data classes and field validation.
Also, pydantic allows us to convert classes to dictionaries.
Here is an example with a basic player class

```python
from pydantic import BaseModel

class Player(BaseModel):
    # The player's name
    name: str
    # The player's age
    age: int
    # The player's health
    health: int = 100
```

This class works like a dataclass. Basically, it only contains data and does not contain functions.
It CAN however contain functions, but the __init__ method will break the functionalities if you are not careful.

Here is an example of us loading JSON and creating a Player class out of it:
```python
# The JSON library
# This is included with python
import json

# Our input string containing json
in_str = """
{
 "name": "playername"
 "age": 15,
 "health: 100
}
"""

# Parse the json
parsed = json.parse(in_str)


# Create the class instance with the data
player_inst = Player(**parsed)
```

Note the ** when we are creating the Player. This takes the dictionary and passes it as keyword arguments to the class.
That is all the time I have for now, so Happy Coding and have a good night.

**Comment by Slow-Scallion4183** [comment]
Thanks for that! Very informative and has been explained well! Examples were also great! Thanks for that :D At some point we will implement this. As you said, it will greatly speed up the development process.
